pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,61 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
 }
 
 // ValidateJobStatusUpdate validates an update to a JobStatus and returns an ErrorList with any errors.
-func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList {
+func ValidateJobStatusUpdate(job, oldJob *batch.Job) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateJobStatus(&status, field.NewPath("status"))...)
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(&job.Status, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// For example, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	// We verify the status fields only if changed or order to allow unrelated
+	// status updates to be possible.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete, batch.JobFailureTarget} {
+			if isConditionTrue(oldJob.Status.Conditions, cType) && !isConditionTrue(job.Status.Conditions, cType) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, fmt.Sprintf("cannot disable the terminal %s=True condition", string(cType))))
+			}
+		}
+		if isConditionTrue(job.Status.Conditions, batch.JobComplete) && isConditionTrue(job.Status.Conditions, batch.JobFailed) {
+			if !isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) || !isConditionTrue(oldJob.Status.Conditions, batch.JobFailed) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, "both Complete=True and Failed=True conditions cannot be set"))
+			}
+		}
+		if oldJob.Status.CompletedIndexes != job.Status.CompletedIndexes {
+			if ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.NonIndexedCompletion && job.Status.CompletedIndexes != "" {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completedIndexes"), job.Status.CompletedIndexes, "cannot set non-empty completedIndexes when non-indexed completion mode"))
+			}
+			if job.Spec.Completions != nil {
+				if err := validateIndexesFormat(job.Status.CompletedIndexes, int(*job.Spec.Completions)); err != nil {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("completedIndexes"), job.Status.CompletedIndexes, fmt.Sprintf("error parsing completedIndexes: %s", err.Error())))
+				}
+			}
+		}
+		if !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) || isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) != isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+			if job.Status.CompletionTime != nil && !isConditionTrue(job.Status.Conditions, batch.JobComplete) {
```

Comment:
Shouldn't we also check that CompletionTime cannot be changed once it's set?

Reply:
I think this makes sense, especially since this affects CronJob. Additionally, we could verify that `CompletionTime` is set whenever the `Complete=True` condition is set. WDYT @alculquicondor ?

Reply:
yes, let's add it.

Reply:
added here: https://github.com/kubernetes/kubernetes/blob/f6c10fadd9e44a9ba319f599613835ce395959f9/pkg/apis/batch/validation/validation.go#L535
marking as resolved, feel free to open a dedicated thread if some improvement is needed.